### ISSUE TITLE
Question about DTO subtypes

### ISSUE THREAD

**Issue by luisfvieirasilva** [oatpp/oatpp#405]
Hi guys,

First of all, thanks for the excellent framework!!
Secondly, is there any way to implement a DTO that have some sub types like [JAVA's Jackson](https://www.baeldung.com/jackson-inheritance#2-per-class-annotations) does?

Copying from the example of the link, imagine that I have a class `Vehicle` and the classes `Car` and `Truck` that inheritance from it, how can I deserialize a json of a Car using `jsonObjectMapper->readFromString<oatpp::Object<Vehicle>>(json)`?

I have search though the serialization/deserialization part of the project's code and imagine this isn't possible, so, if I'm correct, I would appreciate a light of how I can mimic this behavior.

Another difficult that I'm facing with the DTO is releated to the fact that the DTO_INIT macro explicitly defines the default constructor of the class, so I can't initialize the DTO's base class when using it. Am I missing something here?

**Comment by lganzzzo** [comment]
Hello @luisfvieirasilva ,

Let me answer the simple question first:)

> Another difficult that I'm facing with the DTO is releated to the fact that the DTO_INIT macro explicitly defines the default constructor of the class, so I can't initialize the DTO's base class when using it. Am I missing something here?

Actually, it's a good point. I guess we have to remove that default consructor.
However, you still can initialize your DTO:

Initialize fields:

```cpp
class MyDto : public oatpp::DTO {

  DTO_INIT(MyDto, DTO)

  DTO_FIELD(String, str) = "hello";
  DTO_FIELD(Int32, int32) = 32;

};
```

Custom constructor:

```cpp
class MyDto : public oatpp::DTO {

  DTO_INIT(MyDto, DTO)

  DTO_FIELD(String, str);
  DTO_FIELD(Int32, int32);

public:

  MyDto(const String& s, const Int32& i) {
    str = s;
    int32 = i;
  }

};

...

auto dto = MyDto::createShared("hello", 32);
```

---

Now the difficult one:
The inheritance and deserialization...

Currently, there is no way for the deserializer to guess class descendants type from the JSON data - so the short answer, is no, you can't do that.

The workarounds are:
- Declare a "universal" DTO for Vehicle - to contain all possible fields, and then manually pick the correct ones.
- Use `oatpp::Any`
   ```cpp
   oatpp::Any abstractObject = jsonObjectMapper->readFromString<oatpp::Any>(json);
   ```
   While `oatpp::Any` is very convenient for storing data and passing data further to serializers, it's very inconvenient to 
   manipulate `oatpp::Any` when you have to read some exact field from it.

So I would suggest going with the "universal" DTO.

Regards,
Leonid

**Comment by luisfvieirasilva** [comment]
Thank you @lganzzzo for the answer
About the default constructor, I understand that I can initialize my DTO using another constructor, but the problem is that I also need to use the default one that, ideally, need to initialize the DTO base class.

I'm willing to help the project and thought about what is the best way to fix this, once you need to force that the DTO has a default constructor to be used during the deserialization, and found some options:

1. The DTO_INIT macro just declare the default constructor and force the user to define it. The cons is that the user will have to define the constructor outside of the class and all Oat++ users will have to change all its DTO
2. Create another macro similar to DTO_INIT that don't declare the default constructor for this specific case
3. Add a not required third parameter to the DTO_INIT macro for force it don't declare the default constructor, something like `DTO_INIT(MyClass, BaseClass, WITH_CUSTOM_DEFAULT_CONSTRUCTOR)`

In my opinion the the last one is the best although the name WITH_CUSTOM_DEFAULT_CONSTRUCTOR can me improved. I've never contribute to a open source project, so its possible that non of this solution is great, so I'll be glad to hear our opinion and start coding it.

#

Know about the inheritance and deserialization,

I work with a project that uses a bunch of "universal" and big DTOs and I'm traumatized with them because It reaches a point where don't know where and how some fields are used. But you have convince me that the `oatpp::Any` solution isn't the best way either. So I've spend the last days trying to by pass this problem extending the `oatpp::parser::json::mapping::ObjectMapper` and its deserializer. The solution I've found was:

1. My `Vehicle` class have a string "type" field that can have the values "car" or "truck"
2. I have `JsonTypeVerifier` class that has a `oatpp::Type`, a field name and a field value. It has a method that parses a json string (I'm using the [nlohmann](https://github.com/nlohmann/json) json library for its) and returns true if it has the field name with the correct field value.
3. I have a `JsonTypeSelector` that has a list of `JsonTypeVerifier` and a method that given a json string, it returns the `oatpp::Type` of the first `JsonTypeVerifier` that matches the json string
4. I have extend the `oatpp::parser::json::mapping::Deserializer` with the following features:
    1. An unordered map from the DTO`s name to a `JsonTypeSelector` and the methods to add and entry inside it
    2. And override to the `deserialize(oatpp::parser::Caret& caret, const Type* type)` method that checks if it is deserializing an Object and, if the objects name has an entry in the unordered, it uses the `JsonTypeSelector` to get the correct type and calls the original method with it.

The problems I can see with the solution is that I need to parse the json at least twice (one to check and one to create the object) and I'm currently using a third party lib (which I thing is easy to solve). If you and the others project maintainers though this is good feature for the framework I can work on this after we discuss the best syntax for it.

#

Last but not the least, I have some suggestions about two minor things:

1. The `oatpp::parser::json::mapping::Deserializer` has no virtual method and, once I can pass it to the `oatpp::parser::json::mapping::ObjectMapper` constructor, I thing it may be a good thing to allow user to personalize theirs deserialization.
2. I can't deserialize an object and get a unique_ptr of it, once the `ObjectWrapper` class returns me a shared_ptr. I imagine this isn't a simple thing to change but thing forces the user to use shared_ptr where maybe the unique_ptr would be better.

Regards,
Luis

**Comment by lganzzzo** [comment]
Hello @luisfvieirasilva ,

Thank you for the detailed comments, and sorry for the late reply. 

I took a look at the default constructor issue, and I think that the best way will be just to remove the default constructor from `DTO INIT` macro - https://github.com/oatpp/oatpp/blob/master/src/oatpp/codegen/dto/base_define.hpp#L51.
This way I believe it will be the most obvious for the user what happens.

Yes, in this case, when the user wants to define a custom constructor the default constructor is also required.

```cpp

class MyDto : public oatpp::DTO {
  
  DTO_INIT(MyDto, DTO)
  
  DTO_FIELD(String, name) = "Name";
  DTO_FIELD(String, secondName) = "Second Name";

public:

  MyDto() = default; // <-- This guy will also be required.

  MyDto(const char* pName, const char* pSecondName)
    : name(pName)
    , secondName(pSecondName)
  {}

};
```

> Know about the inheritance and deserialization, ...

One of the possible ways to implement inheritance support during JSON deserialization could be the following:

- add additional flags to serialize/deserialize config - something like "config->includeInheritanceInfo = true;"
- automatically add inheritance metadata to serialized DTO
- deserialize taking into account inheritance metadata.

But I'm still thinking about the most efficient way to do it.

> Last but not the least, I have some suggestions about two minor things:

1.
>The oatpp::parser::json::mapping::Deserializer has no virtual method and, once I can pass it to the oatpp::parser::json::mapping::ObjectMapper constructor, I thing it may be a good thing to allow user to personalize theirs deserialization.

You can patch serializer/deserializer by setting ser/de methods for each type:
https://github.com/oatpp/oatpp/blob/master/src/oatpp/parser/json/mapping/Serializer.hpp#L206

 2.
>I can't deserialize an object and get a unique_ptr of it, once the ObjectWrapper class returns me a shared_ptr. I imagine this isn't a simple thing to change but thing forces the user to use shared_ptr where maybe the unique_ptr would be better.

At the moment I don't see how `unique_ptr` can be supported in oatpp object-mapping system.
The whole oatpp object-mapping framework is built around `shared_ptr`. All oatpp types are basically `shared_ptr` to an object plus type information. 
Even if the deserializer could return you an `unique_ptr` to the root object, all of its' fields still would be `shared_ptr`s.

So it's a limitation.

**Comment by lganzzzo** [comment]
Hey @luisfvieirasilva ,

> So I've spend the last days trying to by pass this problem extending the oatpp::parser::json::mapping::ObjectMapper and its deserializer. The solution I've found was:

Actually, your example with `JsonTypeSelector` makes perfect sense. I just have to think about how it can possibly fit inside the oatpp DTO mechanism to make it an oatpp feature.

I'll put the `1.3.0` tag on this issue. If it's possible to make it, I want to have it by `1.3.0` release.

The limitation that I can see right away is that the type-defining field should be inside the DTO itself.
I mean:

```json
{
  "vehicle": {
    "type": "car" // this will work
  }
}
```

```json
{
  "vehicle_type": "car", // this will NOT work
  "vehicle": {}
}
```

What I'm thinking of right now is something like this:

```cpp
class Vehicle : public oatpp::DTO {

  DTO_INIT(Vehicle, DTO)

  DTO_FIELD(Enum<VehicleType>::AsString::NotNull, vehicleType); // type-defining field
  
  
  /* DTO type selector */
  DTO_TYPE_SELECTOR(vehicleType /* multiple fields allowed */) {
    switch(*vehicleType) {
      case VehicleType::CAR: return Car::createShared();
      default:
        return nullptr;
    }
  }

};
```

So deserializer will first look for type-defining fields, and after the type is resolved - will generate the corresponding DTO.

**Comment by lganzzzo** [comment]
`INIT_DTO` default constructor is removed in this commit - https://github.com/oatpp/oatpp/commit/15f77f1ed22819ae834fffe4bd35510544273e2a

**Comment by luisfvieirasilva** [comment]
Hello @lganzzzo,
Thanks for showing to me the `setSerializerMethod`. It's very useful!!

About the deserialization, I have a suggestion that honestly I don't know if it is possible to implement, but I imagine that would looks great for the final user.

```c++
class Car;
class Truck;

class Vehicle : public oatpp::DTO {
    DTO_INIT(Vehicle, DTO)
    DTO_TYPE_DEFINE_FIELD(Enum<VehicleType>::AsString::NotNull, vehicleType);

    Vehicle(VehicleType type) : vehicleType(type) {}

    // Will be better if the following lines wasn't necessary, but I can't see how
    DTO_SUB_TYPE(Car)
    DTO_SUB_TYPE(Truck)
    
}

class Car : public Vehicle {
    SUB_DTO_INIT(Car, Vehicle, VehicleType::CAR)
}

class Truck : public Vehicle {
    SUB_DTO_INIT(Truck, Vehicle, VehicleType::TRUCK)
}

```
I think this way makes more transparent to the final user what your solution achieves with the DTO_TYPE_SELECTOR. Of course this solution has the limitation to force just one field was type defining, but I imagine that this is sufficient.

One another problem that a had during my by pass was that the `oatpp::data::mapping ::type::ObjectWrapper` assign operator don't copy the value type. I conclude that is a bug since I can assign a Void to another one that not necessary has the same type.

Any way, I would like to reinforce my willing to contribute to the project, so after deciding what are the best fit of the deserialization inside the oatpp framework I'll be glad to start coding! Is it OK for you?

**Comment by lganzzzo** [comment]
Hey @luisfvieirasilva ,

> About the deserialization, I have a suggestion that honestly I don't know if it is possible to implement, but I imagine that would looks great for the final user.
...
> I think this way makes more transparent to the final user what your solution achieves with the DTO_TYPE_SELECTOR. Of course this solution has the limitation to force just one field was type defining, but I imagine that this is sufficient

The problem with this solution is that oatpp has no knowledge of how to use the defined `DTO_TYPE_DEFINE_FIELD`, it doesn't know the possible values. 
Also, it requires an additional macro to define `DTO_TYPE_DEFINE_FIELD` which is actually the same `DTO_FIELD` as others - it just that we want to build some "inheritance" functionality using that fields. So I think it's better not to introduce an additional macro for that purpose.

It can possibly look like the following:

```cpp
class Car;
class Truck;

class Vehicle : public oatpp::DTO {

    DTO_INIT(Vehicle, DTO)

    DTO_FIELD(Enum<VehicleType>::AsString::NotNull, vehicleType);

    DTO_SUB_TYPE(
      Car,             // DTO class
      vehicleType,     // type field
      VehicleType::CAR // falue of vehicleType to produce CAT during deserialization
    )
    
    DTO_SUB_TYPE(
      Truck, 
      vehicleType, 
      VehicleType::TRUCK
    )
    
};

class Car : public Vehicle {

  DTO_INIT(Car, Vehicle)

};

class Truck : public Vehicle {

    DTO_INIT(Truck, Vehicle)

};
```

But also I would prefer to use a user-defined method to map DTO type-fields to DTO subclasses.


> One another problem that a had during my by pass was that the oatpp::data::mapping ::type::ObjectWrapper assign operator don't copy the value type. I conclude that is a bug since I can assign a Void to another one that not necessary has the same type.

This is by design. Because when a DTO field created, its' type can't change.
Once an object-wrapper instance is created its type won't be changed - you can only "cast" it to another object-wrapper:

```cpp
oatpp::Void v;

oatpp::String a = "hello";

v = s;

oatpp::String b = v.staticCast<oatpp::String>();

OATPP_LOGD("AAA", "'%s'", b->c_str());
```

> Any way, I would like to reinforce my willing to contribute to the project, so after deciding what are the best fit of the deserialization inside the oatpp framework I'll be glad to start coding! Is it OK for you?

Sure, contributions are very welcome!
The first thing that we need to implement an inheritance feature is to understand how the deserializer can get inheritance information from the DTO and how we are going to store the inheritance info in the DTO.

Ideally is to put that information inside the [DTOs' type](https://github.com/oatpp/oatpp/blob/master/src/oatpp/core/data/mapping/type/Object.hpp#L235) which looks to be achievable by just extending the [polymorphic dispatcher of the abstract object](https://github.com/oatpp/oatpp/blob/master/src/oatpp/core/data/mapping/type/Object.hpp#L177)

Then we have to extend the [DTO codegen](https://github.com/oatpp/oatpp/blob/master/src/oatpp/codegen/dto/base_define.hpp) to add a default (empty) inheritance information plus we should be able to override that info by adding a user-defined method for mapping fields to DTO classes.

After we have that, all we have to do is to edit deserializer methods to support this.
Please also note that oatpp tries to keep the consistency of object-mapping framework across all oatpp features - which means that this inheritance-tracking capability of deserializer should be added everywhere including
- JSON
- BSON
- ORM SQLite deserializer
- ORM Postgres deserializer

**Comment by luisfvieirasilva** [comment]
Hi @lganzzzo 

> The problem with this solution is that oatpp has no knowledge of how to use the defined `DTO_TYPE_DEFINE_FIELD`, it doesn't know the possible values.
> Also, it requires an additional macro to define `DTO_TYPE_DEFINE_FIELD` which is actually the same `DTO_FIELD` as others - it just that we want to build some "inheritance" functionality using that fields. So I think it's better not to introduce an additional macro for that purpose.

The idea was the `DTO_TYPE_DEFINE_FIELD` macro initialize some kind of map that the `SUB_DTO_INIT` macro could use to register the sub type, but the user-defined method looks good too.

> This is by design. Because when a DTO field created, its' type can't change.
> Once an object-wrapper instance is created its type won't be changed - you can only "cast" it to another object-wrapper:

What I couldn't understand about this design is the fact that I can do this for exemple:
```c++
oatpp::String s = "hello";
oatpp::Void v(s);

oatpp::Int32 i = 1;

v = i;
```

At this point `v` has a void pointer to a v_uint32 and the String's type. Here what I see as the main problem it that some constructors allows the user to define at the runtime a type that is differ from the static Class::getType(), but the assign operator can't change it and neither checks if the assignment makes sense. Now considering inheritance classes, since I can assign a pointer of a derivative class to a pointer of a base class, should the object-wrapper's type contains the actual type of its pointer or the type of the base? If it should contains the base type, we'll need to get the correct type before the serialization occurs.

> Sure, contributions are very welcome!

I'll start coding this this weekend! Thank you for your description of how do this and where to look!

Also, I didn't find the code style definition of the project, do you have a clang-format file or something similar?

**Comment by luisfvieirasilva** [comment]
Hey @lganzzzo,
I was thinking about your `DTO_TYPE_SELECTOR`solution and have a question about it. During the deserialization of an object, we first have to deserialize the json to the base class (ignoring extra fields), and then deserialize again with the correct subtype returned by the user defined method, is that correct?

**Comment by lganzzzo** [comment]
Hey @luisfvieirasilva ,

> What I couldn't understand about this design is the fact that I can do this for exemple:

```cpp
oatpp::String s = "hello";
oatpp::Void v(s);

oatpp::Int32 i = 1;

v = I;
```

Maybe you are right, maybe we should make the `ObjectWrapper::valueType` not const, and allow reassignment of a `valueType`. I can't tell at the moment what it can lead to - I have to look at this more closely. I was already thinking to do it in the past but for some reason, I decided that it wasn't a good idea - I don't remember.

> Also, I didn't find the code style definition of the project, do you have a clang-format file or something similar?

There is no code-style file in the repo. Please just mimic the code style of the project.
If there is something too crazy with a code-style in your code - we can always fix that on code-review.

> I was thinking about your DTO_TYPE_SELECTORsolution and have a question about it. During the deserialization of an object, we first have to deserialize the json to the base class (ignoring extra fields), and then deserialize again with the correct subtype returned by the user defined method, is that correct?

We don't want to do deserialization twice.
The idea is the following:
- parser scans JSON object and deserializes each field storing each field in the `std::unordered_map<std::string, oatpp::Void>`
- then we call `DTO_TYPE_SELECTOR` providing already deserialized fields to it (which we have already stored in the map).
- `DTO_TYPE_SELECTOR` creates a correct object.
- We iterate through all fields of the created object assigning the correspondent values from the map.

Thus we don't need to do extra allocations for the base class.

**Comment by lganzzzo** [comment]
Implemented in `1.3.0` see [changelog](https://github.com/oatpp/oatpp/blob/master/changelog/1.3.0.md#polymorphic-dto_field)
